Add meta device to generic device testing framework, skip NotImplementedError

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53973 Add some trivial meta functions for commonly used inplace operations
* #53972 Make storage access error NotImplementedError
* #53823 Compute type_equal() without reference to backend()
* #53816 Also propagate storage_access_should_throw_ when copying tensor metadata
* #53759 Fix restriding logic for structured kernels
* **#53682 Add meta device to generic device testing framework, skip NotImplementedError**
* #53747 Add 'noarch' tests which only run in one CI config

With this, under the meta device, 101 tests passed and 16953 skipped.
It ain't much, but it's a start.

Some various bits and bobs:
- NotImplementedError suppression at test level is implemented
  in the same way as CUDA memory leak check, i.e., by wrapping
  test methods and monkeypatching them back in.
- I had to reimplement assertRaises/assertRaisesRegex from scratch to
  ignore NotImplementedError when _ignore_not_implemented_error is True.
  The implementation relies on a small amount of private API that hasn't
  changed since 2010
- expectedAlertNondeterministic doesn't really work so I skipped them
  all; there's probably a way to do it better

I tested this using `pytest --disable-warnings --tb=native -k meta --sw
test/*.py` and a pile of extra patches to make collection actually work
(lol).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26955539](https://our.internmc.facebook.com/intern/diff/D26955539)

ezyang committed:
Add meta device to generic device testing framework, skip NotImplementedError

With this, under the meta device, 101 tests passed and 16953 skipped.
It ain't much, but it's a start.

Some various bits and bobs:
- NotImplementedError suppression at test level is implemented
  in the same way as CUDA memory leak check, i.e., by wrapping
  test methods and monkeypatching them back in.
- I had to reimplement assertRaises/assertRaisesRegex from scratch to
  ignore NotImplementedError when _ignore_not_implemented_error is True.
  The implementation relies on a small amount of private API that hasn't
  changed since 2010
- expectedAlertNondeterministic doesn't really work so I skipped them
  all; there's probably a way to do it better

I tested this using `pytest --disable-warnings --tb=native -k meta --sw
test/*.py` and a pile of extra patches to make collection actually work
(lol).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
Here is what is currently causing tests to skip:

```
            {'aten::_copy_from': 2167,
             'aten::arange.start_out': 24,
             'aten::as_strided': 8,
             'aten::eye.m_out': 45,
             'aten::fill_.Scalar': 138,
             'aten::linspace.out': 20,
             'aten::logspace.out': 18,
             'aten::normal_': 1757,
             'aten::random_': 9,
             'aten::random_.from': 4103,
             'aten::randperm.generator_out': 20,
             'aten::range.out': 2,
             'aten::take': 1,
             'aten::unfold': 1,
             'aten::uniform_': 4236,
             'aten::view': 4,
             'aten::zero_': 62,
             'torch.tensor': 1146})
```

computed by looking at the junit xml with

```
from junitparser import JUnitXml
import re
from collections import defaultdict
from pprint import pprint

xml = JUnitXml.fromfile('foo.xml')

result = defaultdict(int)

for suite in xml:
    for case in suite:
        if case.result:
            r = case.result[0]
            if r.type == "pytest.skip" and r.message.startswith("not implemented"):
                m = re.search(r"Could not run '([^']+)'", r.text)
                if m:
                    op = m.group(1)
                elif "torch.tensor" in r.text:
                    op = "torch.tensor"
                else:
                    continue
                result[op] += 1

pprint(result)
```

ezyang committed:
Update on "Add meta device to generic device testing framework, skip NotImplementedError"


With this, under the meta device, 101 tests passed and 16953 skipped.
It ain't much, but it's a start.

Some various bits and bobs:
- NotImplementedError suppression at test level is implemented
  in the same way as CUDA memory leak check, i.e., by wrapping
  test methods and monkeypatching them back in.
- I had to reimplement assertRaises/assertRaisesRegex from scratch to
  ignore NotImplementedError when _ignore_not_implemented_error is True.
  The implementation relies on a small amount of private API that hasn't
  changed since 2010
- expectedAlertNondeterministic doesn't really work so I skipped them
  all; there's probably a way to do it better

I tested this using `pytest --disable-warnings --tb=native -k meta --sw
test/*.py` and a pile of extra patches to make collection actually work
(lol).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -2691,12 +2693,14 @@ def _test_linspace(self, device, dtype, steps):
+    @skipMeta  # See https://github.com/pytorch/pytorch/issues/53647
     @skipCPUIf(True, "compares with CPU")
     @precisionOverride({torch.half: 0.0039 + LINSPACE_LOGSPACE_EXTRA_EPS})
     @dtypesIfCUDA(*(torch.testing.get_all_fp_dtypes() + torch.testing.get_all_complex_dtypes()))
```
@dtypes() could be added to this test

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -816,6 +823,14 @@ def get_comparison_dtype(a, b):
 
+# The year is 2021: this private class hierarchy hasn't changed since 2010,
+# seems low risk to inherit from
+class AssertRaisesContextIgnoreNotImplementedError(unittest.case._AssertRaisesContext):
```
This comment could also explain what the class is intended for

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -846,6 +861,7 @@ def precision(self, prec: float) -> None:
 
     _do_cuda_memory_leak_check = False
     _do_cuda_non_default_stream = False
+    _ignore_not_implemented_error = False
```
Suggest adding a comment describing this setting

mruberry commented on test/test_linalg.py:
```diff
@@ -2985,6 +2985,9 @@ def run_test_skipped_elements(a_shape, ind):
         run_test_skipped_elements((12, 3, 2), ind=1)
         run_test_skipped_elements((18, 3, 3, 1), ind=1)
 
+    # TODO: meta fails with "RuntimeError: Failed to invert the input tensor, because
```
Nit: might be easier to track by filing and linking an issue

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1264,6 +1292,26 @@ def assertObjectIn(self, obj: Any, iterable: Iterable[Any]) -> None:
                 return
         raise AssertionError("object not found in iterable")
 
+    def assertRaises(self, expected_exception, *args, **kwargs):
```
Add comments here and for assertRaisesRegex() below on why the override is needed.

I can appreciate why we may need an override, but do we really need to subclass unittest.case._AssertRaisesContext? Seems like the overrides should give us the flexibility not to?

mruberry commented:
Adding a meta device seems cool. How time consuming are meta tests? Would it make sense to limit them to only certain builds? 

I made a few inline comments. Mostly suggestions for additional comments, but a sanity check on subclassing unittest.case._AssertRaisesContext, too.

ezyang commented:
> How time consuming are meta tests? 

On an optimized build of PyTorch, it still takes a few minutes to run them all because Python is slow :( (according to a XML file I have lying around, it took just shy of 7min to run all test with "meta" in their name). 7min isn't half an hour though, so maybe it's OK? I think it would be pretty OK to only run these on one build, though, because (1) a regular CPU kernel will still exercise all this and (2) the meta kernels aren't very platform sensitive (unlike other sense)

mruberry commented:
> > How time consuming are meta tests?
> 
> On an optimized build of PyTorch, it still takes a few minutes to run them all because Python is slow :( (according to a XML file I have lying around, it took just shy of 7min to run all test with "meta" in their name). 7min isn't half an hour though, so maybe it's OK? I think it would be pretty OK to only run these on one build, though, because (1) a regular CPU kernel will still exercise all this and (2) the meta kernels aren't very platform sensitive (unlike other sense)

Limiting them to one build would be preferable if possibly. 7 minutes isn't so bad but 7 minutes on every test build adds up.

I also forgot to mention that some tests use the @onlyOnCPUAndCUDA decorator, but those tests probably want to be @onlyOnCPUAndCUDAAndMeta now. Should we change the decorators so that if CPU is run then Meta is run, too, and require explicitly skipping meta? So @onlyCPU would actually be @onlyCPUAndMeta, for example? And to actually run exclusively on the CPU the test would need to have both the @onlyCPU decorator and the @skipMeta decorator?

ezyang commented:
> I also forgot to mention that some tests use the @onlyOnCPUAndCUDA decorator, but those tests probably want to be @onlyOnCPUAndCUDAAndMeta now. Should we change the decorators so that if CPU is run then Meta is run, too, and require explicitly skipping meta? So @onlyCPU would actually be @onlyCPUAndMeta, for example? And to actually run exclusively on the CPU the test would need to have both the @onlyCPU decorator and the @skipMeta decorator?

I think we need to sit down and have a serious design discussion about how to organize these decorators.

ezyang committed:
Update on "Add meta device to generic device testing framework, skip NotImplementedError"


With this, under the meta device, 101 tests passed and 16953 skipped.
It ain't much, but it's a start.

Some various bits and bobs:
- NotImplementedError suppression at test level is implemented
  in the same way as CUDA memory leak check, i.e., by wrapping
  test methods and monkeypatching them back in.
- I had to reimplement assertRaises/assertRaisesRegex from scratch to
  ignore NotImplementedError when _ignore_not_implemented_error is True.
  The implementation relies on a small amount of private API that hasn't
  changed since 2010
- expectedAlertNondeterministic doesn't really work so I skipped them
  all; there's probably a way to do it better

I tested this using `pytest --disable-warnings --tb=native -k meta --sw
test/*.py` and a pile of extra patches to make collection actually work
(lol).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add meta device to generic device testing framework, skip NotImplementedError"


With this, under the meta device, 101 tests passed and 16953 skipped.
It ain't much, but it's a start.

Some various bits and bobs:
- NotImplementedError suppression at test level is implemented
  in the same way as CUDA memory leak check, i.e., by wrapping
  test methods and monkeypatching them back in.
- I had to reimplement assertRaises/assertRaisesRegex from scratch to
  ignore NotImplementedError when _ignore_not_implemented_error is True.
  The implementation relies on a small amount of private API that hasn't
  changed since 2010
- expectedAlertNondeterministic doesn't really work so I skipped them
  all; there's probably a way to do it better

I tested this using `pytest --disable-warnings --tb=native -k meta --sw
test/*.py` and a pile of extra patches to make collection actually work
(lol).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26955539](https://our.internmc.facebook.com/intern/diff/D26955539)

[ghstack-poisoned]

ezyang committed:
Update on "Add meta device to generic device testing framework, skip NotImplementedError"


With this, under the meta device, 101 tests passed and 16953 skipped.
It ain't much, but it's a start.

Some various bits and bobs:
- NotImplementedError suppression at test level is implemented
  in the same way as CUDA memory leak check, i.e., by wrapping
  test methods and monkeypatching them back in.
- I had to reimplement assertRaises/assertRaisesRegex from scratch to
  ignore NotImplementedError when _ignore_not_implemented_error is True.
  The implementation relies on a small amount of private API that hasn't
  changed since 2010
- expectedAlertNondeterministic doesn't really work so I skipped them
  all; there's probably a way to do it better

I tested this using `pytest --disable-warnings --tb=native -k meta --sw
test/*.py` and a pile of extra patches to make collection actually work
(lol).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26955539](https://our.internmc.facebook.com/intern/diff/D26955539)

[ghstack-poisoned]

ezyang commented:
@mruberry all review comments addressed

ezyang committed:
Update on "Add meta device to generic device testing framework, skip NotImplementedError"


With this, under the meta device, 101 tests passed and 16953 skipped.
It ain't much, but it's a start.

Some various bits and bobs:
- NotImplementedError suppression at test level is implemented
  in the same way as CUDA memory leak check, i.e., by wrapping
  test methods and monkeypatching them back in.
- I had to reimplement assertRaises/assertRaisesRegex from scratch to
  ignore NotImplementedError when _ignore_not_implemented_error is True.
  The implementation relies on a small amount of private API that hasn't
  changed since 2010
- expectedAlertNondeterministic doesn't really work so I skipped them
  all; there's probably a way to do it better

I tested this using `pytest --disable-warnings --tb=native -k meta --sw
test/*.py` and a pile of extra patches to make collection actually work
(lol).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26955539](https://our.internmc.facebook.com/intern/diff/D26955539)

[ghstack-poisoned]

ezyang committed:
Update on "Add meta device to generic device testing framework, skip NotImplementedError"


With this, under the meta device, 101 tests passed and 16953 skipped.
It ain't much, but it's a start.

Some various bits and bobs:
- NotImplementedError suppression at test level is implemented
  in the same way as CUDA memory leak check, i.e., by wrapping
  test methods and monkeypatching them back in.
- I had to reimplement assertRaises/assertRaisesRegex from scratch to
  ignore NotImplementedError when _ignore_not_implemented_error is True.
  The implementation relies on a small amount of private API that hasn't
  changed since 2010
- expectedAlertNondeterministic doesn't really work so I skipped them
  all; there's probably a way to do it better

I tested this using `pytest --disable-warnings --tb=native -k meta --sw
test/*.py` and a pile of extra patches to make collection actually work
(lol).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26955539](https://our.internmc.facebook.com/intern/diff/D26955539)

[ghstack-poisoned]

mruberry commented:
Offline discussion with @ezyang is tracked in #53849.

mruberry reviewed (approved):
LGTM!

ezyang committed:
Update on "Add meta device to generic device testing framework, skip NotImplementedError"


With this, under the meta device, 101 tests passed and 16953 skipped.
It ain't much, but it's a start.

Some various bits and bobs:
- NotImplementedError suppression at test level is implemented
  in the same way as CUDA memory leak check, i.e., by wrapping
  test methods and monkeypatching them back in.
- I had to reimplement assertRaises/assertRaisesRegex from scratch to
  ignore NotImplementedError when _ignore_not_implemented_error is True.
  The implementation relies on a small amount of private API that hasn't
  changed since 2010
- expectedAlertNondeterministic doesn't really work so I skipped them
  all; there's probably a way to do it better

I tested this using `pytest --disable-warnings --tb=native -k meta --sw
test/*.py` and a pile of extra patches to make collection actually work
(lol).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26955539](https://our.internmc.facebook.com/intern/diff/D26955539)

[ghstack-poisoned]

ezyang committed:
Update on "Add meta device to generic device testing framework, skip NotImplementedError"


With this, under the meta device, 101 tests passed and 16953 skipped.
It ain't much, but it's a start.

Some various bits and bobs:
- NotImplementedError suppression at test level is implemented
  in the same way as CUDA memory leak check, i.e., by wrapping
  test methods and monkeypatching them back in.
- I had to reimplement assertRaises/assertRaisesRegex from scratch to
  ignore NotImplementedError when _ignore_not_implemented_error is True.
  The implementation relies on a small amount of private API that hasn't
  changed since 2010
- expectedAlertNondeterministic doesn't really work so I skipped them
  all; there's probably a way to do it better

I tested this using `pytest --disable-warnings --tb=native -k meta --sw
test/*.py` and a pile of extra patches to make collection actually work
(lol).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26955539](https://our.internmc.facebook.com/intern/diff/D26955539)

[ghstack-poisoned]

ezyang committed:
Update on "Add meta device to generic device testing framework, skip NotImplementedError"


With this, under the meta device, 101 tests passed and 16953 skipped.
It ain't much, but it's a start.

Some various bits and bobs:
- NotImplementedError suppression at test level is implemented
  in the same way as CUDA memory leak check, i.e., by wrapping
  test methods and monkeypatching them back in.
- I had to reimplement assertRaises/assertRaisesRegex from scratch to
  ignore NotImplementedError when _ignore_not_implemented_error is True.
  The implementation relies on a small amount of private API that hasn't
  changed since 2010
- expectedAlertNondeterministic doesn't really work so I skipped them
  all; there's probably a way to do it better

I tested this using `pytest --disable-warnings --tb=native -k meta --sw
test/*.py` and a pile of extra patches to make collection actually work
(lol).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26955539](https://our.internmc.facebook.com/intern/diff/D26955539)

[ghstack-poisoned]
